examples/server/server.cpp

```diff
@@ -111,9 +117,603 @@ struct slot_params {
+    llama_token tok;
+    std::string text_to_send;
+    struct token_prob {
+        llama_token tok;
+        std::string tok_str;
+        float prob;
+    };
+    std::vector<token_prob> probs;
+
+    json to_json() const {
+        json probs_for_token = json::array();
+        for (const auto & p : probs) {
+            probs_for_token.push_back(json {
+                {"tok_str", p.tok_str},
+                {"prob",    p.prob},
+            });
+        }
+        return probs_for_token;
+    }
+
+    static json probs_vector_to_json(const std::vector<completion_token_output> & probs) {
+        json out = json::array();
+        for (const auto & prob : probs) {
+            const std::string tok_str = prob.text_to_send;
+            out.push_back(json {
+                {"content", tok_str},
+                {"probs",   prob.to_json()},
+            });
+        }
+        return out;
+    }
+};
+
+struct server_task_result_cmpl_final : server_task_result {
+    int index = 0;
+    std::string content;
+    bool stream;
+    result_timings timings;
+    std::string prompt;
+
+    bool truncated;
+    int32_t n_decoded;
+    int32_t n_prompt_tokens;
+    int32_t n_tokens_cached;
+    int32_t has_new_line;
+    std::string stopping_word;
+    stop_type stop = STOP_TYPE_NONE;
+
+    std::vector<completion_token_output> probs_output;
+
+    slot_params generation_params;
+
+    // OAI-compat fields
+    bool        verbose        = false;
+    bool        oaicompat      = false;
+    bool        oaicompat_chat = true; // TODO: support oaicompat for non-chat
+    std::string oaicompat_model;
+    std::string oaicompat_cmpl_id;
+
+    virtual int get_index() override {
+        return index;
+    }
+
+    virtual json to_json() override {
+        if (oaicompat) {
+            return to_json_oai_compat();
+        }
+        // otherwise, non-OAI-compat JSON
+        json res = json {
+            {"index",               index},
+            {"content",             content},
+            {"id_slot",             id_slot},
+            {"stop",                true},
+            {"model",               oaicompat_model},
+            {"tokens_predicted",    n_decoded},
+            {"tokens_evaluated",    n_prompt_tokens},
+            {"generation_settings", generation_params.to_json()},
+            {"prompt",              prompt},
+            {"has_new_line",        has_new_line},
+            {"truncated",           truncated},
+            {"stop_type",           stop_type_to_str(stop)},
+            {"stopping_word",       stopping_word},
+            {"tokens_cached",       n_tokens_cached},
+            {"timings",             timings.to_json()},
+        };
+        if (!probs_output.empty()) {
+            res["completion_probabilities"] = completion_token_output::probs_vector_to_json(probs_output);
+        }
+        return res;
+    }
+
+    json to_json_oai_compat() {
+        std::string finish_reason = "length";
```

Comment:
@Nero7991 I'm gonna merge this PR soon. You can adapt your PR #10645 to take advantage of this `to_json_oai_compat()`. Please note that for `/completion` endpoint, `oaicompat_chat` will be false.

Since this is a private function, you can even refactor this function into `to_json_oaicompat` and `to_json_oaicompat_completion` for these 2 different cases, then have an if..else branch in `to_json` to select the correct one